src/main/java/org/ow2/proactive/connector/iaas/service/InstanceService.java

```diff
@@ -42,37 +43,72 @@
+                                                                                          " does not exists"));
 
+        Optional.ofNullable(infrastructureService.getInfrastructure(infrastructureId))
+                .ifPresent(infrastructure -> instanceCache.registerInfrastructureInstances(infrastructure,
+                                                                                           instancesCreated));
+
+        return instancesCreated;
     }
 
-    public void deleteInstance(String infrastructureId, String instanceId) {
-        Optional.ofNullable(infrastructureService.getInfrastructure(infrastructureId))
-                .ifPresent(infrastructure -> cloudManager.deleteInstance(infrastructure, instanceId));
+    public void deleteCreatedInstances(String infrastructureId) {
+        Optional.ofNullable(infrastructureService.getInfrastructure(infrastructureId)).ifPresent(infrastructure -> {
+            instanceCache.getCreatedInstances()
+                         .get(infrastructure.getId())
+                         .forEach(instance -> cloudManager.deleteInstance(infrastructure, instance.getId()));
+            instanceCache.deleteAllInfrastructureInstances(infrastructure);
+        });
+    }
 
+    public void deleteInstance(String infrastructureId, String instanceId) {
+        Optional.ofNullable(infrastructureService.getInfrastructure(infrastructureId)).ifPresent(infrastructure -> {
+            cloudManager.deleteInstance(infrastructure, instanceId);
+            instanceCache.getCreatedInstances()
+                         .get(infrastructure.getId())
+                         .stream()
+                         .filter(instance -> instance.getId().equals(instanceId))
+                         .findAny()
+                         .ifPresent(instance -> instanceCache.deleteInfrastructureInstance(infrastructure, instance));
+        });
     }
 
     public void deleteInstanceByTag(String infrastructureId, String instanceTag) {
-        Set<Instance> instances = getAllInstances(infrastructureId);
-        instances.stream()
-                 .filter(instance -> instance.getTag().equals(instanceTag))
-                 .forEach(instance -> cloudManager.deleteInstance(infrastructureService.getInfrastructure(infrastructureId),
-                                                                  instance.getId()));
+        getAllInstances(infrastructureId).stream()
+                                         .filter(instance -> instance.getTag().equals(instanceTag))
+                                         .forEach(instance -> {
+                                             Infrastructure infrastructure = infrastructureService.getInfrastructure(infrastructureId);
+                                             cloudManager.deleteInstance(infrastructure, instance.getId());
+                                             instanceCache.getCreatedInstances()
```

Comment:
better to move this logic into the instanceCache. so here you can just do instanceCache.remove(infrastructureId, instance.getId()).